This report contains, in addition to the code provided by the lecturer, my own code changes and additions, and written answers to answer coursework 10 in Machine Learning and Statistics.

In [1]:
# Load the libraries needed
%matplotlib inline
import csv
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression, Ridge, Lasso
import time

# Seed random number generator
np.random.seed(123)

# Helper functions for linear models
def linear ( X, coefficients ):
    return np.dot(X, np.transpose(coefficients), )

def fit_linear( X, y, features=None ):
    """ 
        Returns the coefficients of a linear model fit to X,y.
        If features is a list of integers, then fit will ignore
        any features whose index is not in the list.
        ( Returned coefficients for these features will be set
        to 0. )
    """
    if features is not None:
        # Make a mask
        tot_num_features = np.shape(X)[-1]
        mask = np.zeros((1,tot_num_features))
        mask[0,features] = 1.
        # Zero out all irrellevant features
        X = X * mask
    
    # Do linear least squares fit
    clf = LinearRegression(fit_intercept=False)
    clf.fit(X, y)
    return clf.coef_


def ridge_regression( X, y, lam=1.0, features=None ):
    """ 
        Identical to fit_linear, but performs ridge regression
        with weight penalty alpha (alternatively known as lambda)
    """
    if features is not None:
        # Make a mask
        tot_num_features = np.shape(X)[-1]
        mask = np.zeros((1,tot_num_features))
        mask[0,features] = 1.
        # Zero out all irrellevant features
        X = X * mask
    
    # Do ridge regression fit
    clf = Ridge(alpha=lam, fit_intercept=False)
    clf.fit(X, y)
    return clf.coef_

def lasso_regression( X, y, lam=1.0, features=None ):
    """ 
        Identical to fit_linear, but performs lasso regression
        with weight penalty alpha (alternatively known as lambda)
    """
    if features is not None:
        # Make a mask
        tot_num_features = np.shape(X)[-1]
        mask = np.zeros((1,tot_num_features))
        mask[0,features] = 1.
        # Zero out all irrellevant features
        X = X * mask
    
    # Do ridge regression fit
    clf = Lasso(alpha=lam, fit_intercept=False, max_iter=1e5)
    clf.fit(X, y)
    return clf.coef_

In [2]:
# MSE score
# You should modify the function mse_loss to compute
# the Mean Square Error given a dataset X, y, and learned
# linear model with coefficients learned_coeff.

def mse_loss( X, y, learned_coefficients ):
    y_pred = linear(X, learned_coefficients)
    rss = np.sum(np.square(y_pred - y))
    mse = rss / len(X)
    return mse

In [3]:
# Fitting linear models with num_examples > num_features
# Run the following code to see the Mean Square Error of three
# linear models fit with Linear Regression, Ridge Regression, and
# LASSO when trained on a dataset with 100 examples of 50 features
# each. You should add in the appropriate code to calculate the test
# error of each trained model on the provided test set X_test,
# y_test .

# Load generated data
X, [y] = np.load("./class10_training_a.npy")
X_test, [y_test] = np.load("./class10_test.npy")

lin_reg_train_loss = mse_loss(X, y, fit_linear(X, y))
ridge_train_loss = mse_loss(X, y, ridge_regression(X, y, lam=1.0))
lasso_train_loss = mse_loss(X, y, lasso_regression(X, y, lam=0.003))

print("Train Loss ({:d} datapoints)".format(len(X)))
print("Linear Regression:    {:.3f}".format(lin_reg_train_loss))
print("Ridge Regression:     {:.3f}".format(ridge_train_loss))
print("LASSO:                {:.3f}".format(lasso_train_loss))

print("\nTest Loss ({:d} datapoints)".format(len(X)))

# Include your code for calculating the loss on the test set (X_test, y_test)
# (And remember NEVER to train on the test set)

lin_reg_test_loss = mse_loss(X_test, y_test, fit_linear(X, y))
ridge_test_loss = mse_loss(X_test, y_test, ridge_regression(X, y, lam=1.0))
lasso_test_loss = mse_loss(X_test, y_test, lasso_regression(X, y, lam=0.003))

print("Linear Regression:    {:.3f}".format(lin_reg_test_loss))
print("Ridge Regression:     {:.3f}".format(ridge_test_loss))
print("LASSO:                {:.3f}".format(lasso_test_loss))


Train Loss (100 datapoints)
Linear Regression:    0.057
Ridge Regression:     0.057
LASSO:                0.058

Test Loss (100 datapoints)
Linear Regression:    0.169
Ridge Regression:     0.167
LASSO:                0.158


In [4]:
# Fitting linear models with num_features > num_examples
# Similar to the above, run the following code to see the train error
# of Ridge regression and LASSO, but now trained on a dataset with
# only 25 examples. As before, you should add in the appropriate code
# to calculate the test error of each trained model on test set (X_test,
# y_test) given in the previous code-block.

# Load generated data
X, [y] = np.load("./class10_training_b.npy")

ridge_train_loss = mse_loss(X, y, ridge_regression(X, y, lam=10))
lasso_train_loss = mse_loss(X, y, lasso_regression(X, y, lam=0.1))

print("Train Loss ({:d} datapoints)".format(len(X)))
# Why do we not use Linear Regression?
print("Ridge Regression:     {:.3f}".format(ridge_train_loss))
print("LASSO:                {:.3f}".format(lasso_train_loss))

print("\nTest Loss ({:d} datapoints)".format(len(X)))
# Include your code for calculating the loss on the test set (X_test, y_test)
ridge_test_loss = mse_loss(X_test, y_test, ridge_regression(X, y, lam=10))
lasso_test_loss = mse_loss(X_test, y_test, lasso_regression(X, y, lam=0.1))

print("Ridge Regression:     {:.3f}".format(ridge_test_loss))
print("LASSO:                {:.3f}".format(lasso_test_loss))

Train Loss (25 datapoints)
Ridge Regression:     0.047
LASSO:                0.146

Test Loss (25 datapoints)
Ridge Regression:     0.715
LASSO:                0.655


# Assignment 1

Add your code on calculating the Mean Squared Error (MSE) to the provided code. <br>


(1) Run the provided code (where the hyper-parameter λ is given) to train 3 different linear models
on the data set, class10_training_a, with the OLS, ridge regression and LASSO learning
algorithms, respectively. Then, run your modified code to calculate the MSE on the test set,
class10_test, with 3 trained linear models, respectively. <br>

(2) Run the provided code (where the hyper-parameter λ is given) to train 2 regularised linear
models on the data set, class10_training_b, with ridge regression and LASSO learning
algorithms, respectively. Then, run your modified code to calculate the MSE on the test set,
class10_test, with 2 trained regularised linear regression models, respectively. Comment
on why the OLS cannot be applied to this training data set. <br>

Answer: <br>

See code above for calculating MSE and results on class10_training_a and class10_training_b. <br>

The reason OLS cannot be applied to the second training dataset (class10_training_b) is because the number of observations is smaller than the number of features. When this is the case there is no longer a unique solution to the least squares coefficient estimate, instead there are infinitely many solutions. The model we train would fit the trainingset very well, but will not be able to generalize on new observations. I.e. the model would be overfit as it is too flexible for the small amount of data. When we use Ridge or Lasso regression on the other hand we introduce constraints which can allow us to find a unique solution and by reducing flexibility makes overfitting less likely. 

In [6]:
# Automatic feature selection with LASSO
# Similar to the above, run the following code to see the train error
# Run this code to fit a linear model to a subset of the Automobile
# dataset from the previous class. As before, you should calculate the
# test error of the trained model. Also observe the weights learned via
# LASSO - how do they compare 

# Load data
X_train, [y_train] = np.load("./class10_auto_train.npy")
X_test, [y_test] = np.load("./class10_auto_test.npy")

lam = 0.5
learned_features = lasso_regression(X_train, y_train, lam=lam)
nonzero_features = np.argwhere(~ np.isclose(learned_features, 0.)).squeeze()

print("Fitted LASSO with lambda {:.3f}, learned parameters:".format(lam))
print(learned_features)
print("Non-zero features : {} ({:d} total)".format(list(nonzero_features), 
                                                   len(nonzero_features)))

train_loss = mse_loss(X_train, y_train, learned_features)
print("Train error: {:.3f}".format(train_loss))

test_loss = mse_loss(X_test, y_test, learned_features)
print("Test error:  {:.3f}".format(test_loss))

Fitted LASSO with lambda 0.500, learned parameters:
[ 0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0. -0.]
Non-zero features : [] (0 total)
Train error: 0.949
Test error:  1.101


# Assignment 2

Run the provided code to train LASSO on the training subset, class10_auto_train,
with λ = 0.05, 0.5, respectively. Calculate the MSE on the test subset, class10_auto_test, with
those trained LASSO models with 2 different λ values, respectively. Based on your observation,
comment on the non-zero features achieved by LASSO when different λ-values are used. <br>

Answer: <br>

 - MSE score with lambda = 0.05: Train error = 0.464. Test error = 0.517. Non-zero features : [0, 1, 3, 12]

 - MSE score with lambda 0.50: Train error = 0.949. Test error = 1.101. Non-zero features : []

By testing with many different lambdas of inceasing value in the code above, including 0.05 and 0.5 we see that as lambda increases, the number of selected non-zero features are reduced. And at lambda = 0.5 we are left with no selected features, as all coefficients are set to zero. This happens because as the penaly term becomes larger with an increasing lamba, the coefficients must become smaller to minimize the error - in the end they all become zero.

In [18]:
# Similar to the above, run the following code to see the train error
# Now implement an appropriate method for choosing the parameter lam
# for LASSO on the full automobile dataset. (Code for loading the
# dataset is given below).

# Load data
data = []
continuous_features = [ 0, 1, 9, 10, 11, 12, 13, 16, 18, 19, 20, 21, 22, 23, 24, 25 ]

# Original data is from https://archive.ics.uci.edu/ml/datasets/automobile
with open('./automobile.csv', 'r') as csvfile:
    csvreader = csv.reader(csvfile, delimiter=',', quotechar='\"')
    for row in csvreader:
        try:
            # Get all continuous rows
            data.append([float(row[i]) for i in continuous_features])
        except:
            continue # skip this row since data-processing failed

data = np.array(data)
y = data[:, 0] # target is first value
X = data[:, 1:] # training data is the rest

# Normalize the data to zero mean and unit std
X = (X - np.mean(X, axis=0, keepdims=True)) / np.std(X, axis=0, keepdims=True)
y = (y - np.mean(y)) / np.std(y)

print("Successfully loaded {:d} entries.\n".format(len(X)))

# Implement your method for selecting an appropriate lambda below:
# Selected method is K-fold Cross Validation (tested with K=5 and K=10)

# Shuffle data
sh = list(range(len(X)))
np.random.shuffle(sh)
X = X[sh]
y = y[sh]

# Helper function for getting folds
def get_fold(X, y, fold, num_folds):
    folds_X = np.array_split(X, num_folds)
    test_X = folds_X.pop(fold)
    train_X = np.concatenate(folds_X)
    
    folds_y = np.array_split(y, num_folds)
    test_y = folds_y.pop(fold)
    train_y = np.concatenate(folds_y)
    return train_X, train_y, test_X, test_y

print("Running cross validation with different values of lambda..")

num_folds = 5  #number of folds we split into
lam = [0.001, 0.003, 0.01, 0.03, 0.1,0.3,1] #lambdas we wanto to try
sum_loss= []  #list to hold the sum of the mean losses for each lambda

for l in lam:
    losses = []
    print("Lambda = " + str(l))
    for i in range(num_folds):
        #Get data for fold
        train_X, train_y, test_X, test_y = get_fold(X, y, i, num_folds)

        #Fit model
        learned_coefficients = lasso_regression(train_X, train_y, lam=l)
        
        #Calculate loss
        mean_fold_loss = mse_loss(test_X, test_y, learned_coefficients)
        
        losses.append(mean_fold_loss)

        print("Mean loss for fold {:d}/{:d}: {:.5f}".format(i+1, num_folds, 
                                                            mean_fold_loss))
    
    sum_loss.append(sum(losses))

best_lambda = lam[np.argmin(sum_loss)]
print(sum_loss)

print('Best lambda: {}'.format(best_lambda))

Successfully loaded 160 entries.

Running cross validation with different values of lambda..
Lambda = 0.001
Mean loss for fold 1/5: 0.73129
Mean loss for fold 2/5: 0.44979
Mean loss for fold 3/5: 0.41770
Mean loss for fold 4/5: 0.46689
Mean loss for fold 5/5: 0.54546
Lambda = 0.003
Mean loss for fold 1/5: 0.71238
Mean loss for fold 2/5: 0.43070
Mean loss for fold 3/5: 0.41820
Mean loss for fold 4/5: 0.46079
Mean loss for fold 5/5: 0.55815
Lambda = 0.01
Mean loss for fold 1/5: 0.65691
Mean loss for fold 2/5: 0.38350
Mean loss for fold 3/5: 0.42651
Mean loss for fold 4/5: 0.45954
Mean loss for fold 5/5: 0.56683
Lambda = 0.03
Mean loss for fold 1/5: 0.59008
Mean loss for fold 2/5: 0.37382
Mean loss for fold 3/5: 0.42886
Mean loss for fold 4/5: 0.44656
Mean loss for fold 5/5: 0.57825
Lambda = 0.1
Mean loss for fold 1/5: 0.61988
Mean loss for fold 2/5: 0.52197
Mean loss for fold 3/5: 0.43258
Mean loss for fold 4/5: 0.48690
Mean loss for fold 5/5: 0.56246
Lambda = 0.3
Mean loss for fold 1/5:

In [19]:
#After finding the best lambda (0.03) we can do fit with all the training data

lam = best_lambda
learned_features = lasso_regression(X, y, lam=lam)
nonzero_features = np.argwhere(~ np.isclose(learned_features, 0.)).squeeze()

print("Fitted LASSO with lambda {:.3f}, learned parameters:".format(lam))
print(learned_features)
print("Non-zero features : {} ({:d} total)".format(list(nonzero_features), 
                                                   len(nonzero_features)))

train_loss = mse_loss(X, y, learned_features)
print("Train error: {:.3f}".format(train_loss))


Fitted LASSO with lambda 0.030, learned parameters:
[ 0.41229806 -0.67706057  0.          0.20655253  0.          0.
  0.         -0.          0.          0.03608566  0.06103328 -0.00820271
 -0.          0.          0.        ]
Non-zero features : [0, 1, 3, 9, 10, 11] (6 total)
Train error: 0.444


# Assignment 3

Apply a proper method learnt in Lectures 8-10 to find out the optimal λ value used
in the LASSO learning on the Automobile data set from λ = 0.001, 0.003, 0.01, 0.03, 0.1, 0.3, and
1.0. Comment on the non-zero features obtained by the LASSO trained with the optimal λ value
by comparing them to those achieved in Class 9. It is essential to justify why the method you use
is appropriate to find out the optimal λ value. <br>

Answer: <br>

See code above for implementation and output. Chose to do a K-fold Cross validation to find the optimal lambda in accordance with (1) and what was adviced in the lecture. The reason for using cross-validation to tune the hyper parameter lambda is that we want to choose a lambda which avoids overfitting by finding the lambda which generalise well across the samples which it learns from in each fold. Thus - the lambda which gave the lowest sum mse across folds was the one chosen. <br>


Tried both 5-fold and 10-fold cross validation and got the same best lambda = 0.03 in both cases (the output above is K=5 to avoid a very long report). The selected features using lambda = 0.03 when fit on the entire training dataset is [0, 1, 3, 9, 10, 11] which gives an error of 0.444. These are not the same as when we used forward and backward stepwise feature selection in coursework 9, using AICS-score as criterion. This strategy selected the subset feature [0, 1, 3]. These are included in the results given by the LASSO regression, but we also get 3 additional features. We do see that the coefficient of the additional features this week (9, 10, 11) are relatively much closer to zero to the other three. This might indicate that the backward stepwise feature selection we ran last week rightfully chose the features which explains the data best. <br>

# References

(1) Chapter 6.2.3 in An Introduction to Statistical Learning with Applications in R. By Gareth James, Daniela Witten, Trevor Hastie and Robert Tibshirani. <br>